rinex/src/clock/record.rs

```diff
@@ -0,0 +1,645 @@
+#[derive(Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Clone, EnumString)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum ClockProfileType {
+    /// Data analysis results for receiver clocks
+    /// derived from a set of network receivers and satellites
+    AR,
+    /// Data analysis results for satellites clocks
+    /// derived from a set of network receivers and satellites
+    AS,
+    /// Calibration measurements for a single GNSS receiver
+    CR,
+    /// Discontinuity measurements for a single GNSS receiver
+    DR,
+    /// Broadcast SV clocks monitor measurements
+    MS,
+}
+
+impl std::fmt::Display for ClockProfileType {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::AR => f.write_str("AR"),
+            Self::AS => f.write_str("AS"),
+            Self::CR => f.write_str("CR"),
+            Self::DR => f.write_str("DR"),
+            Self::MS => f.write_str("MS"),
+        }
+    }
+}
+
+/// Clock RINEX record content.
+pub type Record = BTreeMap<Epoch, BTreeMap<ClockKey, ClockProfile>>;
+
+pub(crate) fn is_new_epoch(line: &str) -> bool {
+    // first 2 bytes match a ClockProfileType code
+    if line.len() < 3 {
+        false
+    } else {
+        let content = line.split_at(2).0;
+        ClockProfileType::from_str(content).is_ok()
+    }
+}
+
+/// Builds `RINEX` record entry for `Clocks` data files.   
+/// Returns identified `epoch` to sort data efficiently.  
+/// Returns 2D data as described in `record` definition
+pub(crate) fn parse_epoch(
+    version: Version,
+    content: &str,
+) -> Result<(Epoch, ClockKey, ClockProfile), Error> {
+    let mut lines = content.lines();
+    let line = lines.next().unwrap();
+    const LIMIT: Version = Version { major: 3, minor: 4 };
+    let (dtype, mut rem) = line.split_at(3);
+    let profile_type = ClockProfileType::from_str(dtype.trim())?;
+
+    let clock_type = match version < LIMIT {
+        true => {
+            // old revision
+            let (system_str, r) = rem.split_at(5);
+            rem = r;
+            if let Ok(svnn) = SV::from_str(system_str.trim()) {
+                ClockType::SV(svnn)
+            } else {
+                ClockType::Station(system_str.trim().to_string())
+            }
+        },
+        false => {
+            // modern revision
+            let (system_str, r) = rem.split_at(4);
+            if let Ok(svnn) = SV::from_str(system_str.trim()) {
+                let (_, r) = r.split_at(6);
+                rem = r;
+                ClockType::SV(svnn)
+            } else {
+                let mut content = system_str.to_owned();
+                let (remainder, r) = r.split_at(6);
+                rem = r;
+                content.push_str(remainder);
+                ClockType::Station(content.trim().to_string())
+            }
+        },
+    };
+
+    // Epoch
+    let offset = 4+1 // Y always a 4 digit number, even on RINEX2
```

Comment:
One cute way to do this is e.g. `let offset = "yyyy.MM.dd.hh.mm.sssssssssss".len();`.